Prototyping/Domains/ThemePark/Invasion/acServer.js

```diff
@@ -0,0 +1,855 @@
+                });
+            }
+        },
+        exists: function() {
+            return this.entity !== null;
+        }
+    };
+    return TrackableEntity;
+})();
+
+var PowerSource = (function() {
+    function PowerSource(name, position) {
+        this.name = name;
+        this.position = position;
+        this.health = 1.0;
+        this.trackableEntity = null;
+    }
+
+    PowerSource.prototype = {
+        trackableEntity: null,
+        health: null,
+        rez: function(device) {
+            var newProperties = JSON.parse(JSON.stringify(POWER_SOURCE_PROPERTIES));
+            newProperties.position = this.position;
+            this.trackableEntity = new TrackableEntity(newProperties, "PowerSource_" + this.name);
+        },
+        isAlive: function() {
+            return this.health > invasionUtils.MINIMUM_POWER_DEVICE_HEALTH;
+        },
+        applyDamage: function(damage) {
+            var newHealth = this.health - damage;
+            if (newHealth < invasionUtils.MINIMUM_POWER_DEVICE_HEALTH) {
+                newHealth = invasionUtils.MINIMUM_POWER_DEVICE_HEALTH;
+            }
+            if (newHealth !== this.health) {
+                this.health = newHealth;
+
+                invasionUtils.setPowerDeviceHealth(this.trackableEntity.entity, this.health);
+                
+                Entities.editEntity(this.trackableEntity.entity, {
+                    angularVelocity: Vec3.multiply(POWER_SOURCE_FULL_ANGULAR_VELOCITY, this.health)
+                });
+            }
+            return this.isAlive();
+        },
+        repair: function(addedHealth) {
+            var newHealth = this.health + addedHealth;
+            if (newHealth > invasionUtils.MAXIMUM_POWER_DEVICE_HEALTH) {
+                newHealth = invasionUtils.MAXIMUM_POWER_DEVICE_HEALTH;
+            }
+            if (newHealth !== this.health) {
+                this.health = newHealth;
+
+                invasionUtils.setPowerDeviceHealth(this.trackableEntity.entity, this.health);
+                
+                Entities.editEntity(this.trackableEntity.entity, {
+                    angularVelocity: Vec3.multiply(POWER_SOURCE_FULL_ANGULAR_VELOCITY, this.health)
+                });
+            }
+        }
+    };
+    return PowerSource;
+})();
+
+var UFO_PATH_ACTION = {
+    PAUSE: 0,
+    MOVE: 1,
+    SPIN_MOVE: 2,
+    WARP: 3
+};
+
+var UFOMovePathEvent = (function() {
+    function UFOMovePathEvent(ufoMoveAction, options) {
+        this.ufoMoveAction = ufoMoveAction;
+        this.options = options;
+    }
+
+    UFOMovePathEvent.prototype = {
+        ufoMoveAction: null,
+        options: null
+    };
+    
+    return UFOMovePathEvent;
+})();
+
+var Alien = (function() {
+    function Alien(motherShip, position, targetPowerSource) {
+        this.motherShip = motherShip;
+        this.position = position;
+        var newProperties = JSON.parse(JSON.stringify(ALIEN_PROPERTIES));
```

Comment:
Double parse/stringify?

Reply:
This is the clone technique. Otherwise you keep on changing the contant, which should not be changed really since its a constant.